test/nodes/test_reader.py

```diff
@@ -46,6 +46,74 @@ def test_reader_basic(reader):
     assert isinstance(reader, BaseReader)
 
 
+def test_add_answer_page_number() -> None:
+    documents = [
+        Document(content="This is a test.\fSentence on second page about nothing.", meta={"page_number": 1}),
+        Document(content="Second sentence on the second page.", meta={"page_number": 2}),
+    ]
+    with patch("haystack.nodes.FARMReader.__init__") as mock_reader_init:
+        mock_reader_init.return_value = None
+        reader = FARMReader(model_name_or_path="fake_model", use_gpu=False)
+    answer_with_meta = reader._add_answer_page_number(
+        documents=documents,
+        answer=Answer(
+            answer="nothing",
+            type="extractive",
+            score=0.2,
+            context=documents[0].content,
+            document_ids=[documents[0].id],
+            offsets_in_document=[Span(start=46, end=46 + len("nothing"))],
+        ),
+    )
+    assert answer_with_meta.meta is not None
+    assert answer_with_meta.meta["answer_page_number"] == 2
+
+
+def test_add_answer_page_number_no_doc_page() -> None:
+    documents = [
+        Document(content="This is a test.\fSentence on second page about nothing."),
+        Document(content="Second sentence on the second page."),
+    ]
+    with patch("haystack.nodes.FARMReader.__init__") as mock_reader_init:
+        mock_reader_init.return_value = None
+        reader = FARMReader(model_name_or_path="fake_model", use_gpu=False)
```

Comment:
This is a really strange approach, I think it would be better to mock `initialize_device_settings`, `resolve_torch_dtype` and `QAInferencer`. It's much clearer what you're trying to do, took me some time to understand what was going on. 

It's really not a great practice mocking the class your trying to test, if you make a bad mock your risk testing the mock.

Also to nitpick a bit `__init__` doesn't actually return anything so this is redundant `mock_reader_init.return_value = None`.

Reply:
Same thing on the test above obviously.

Reply:
The `mock_reader_init.return_value = None` is needed otherwise you get 
```python
FAILED test/nodes/test_reader.py::test_add_answer_page_number - TypeError: __init__() should return None, not 'MagicMock'
```

Also I would like to individually test the function `_add_answer_page_number` which doesn't require anything class variables to run, which is why I mocked the init method. So I can make these changes, but would it be better just to make this method static so I can skip this mocking all together?

Reply:
updated

Reply:
Ah right! I forgot that patched calls return a `MagicMock` by default. 🤦 

Making the function static is an option too. Nonetheless the current strategy is good for me. 👍